src/Config.php

```diff
@@ -97,13 +105,25 @@ public function __construct(Composer $composer, string $basePath)
         $extra = (array)($composer->getPackage()->getExtra()[self::CONFIG_KEY] ?? []);
         $this->composerConfig = $composer->getConfig();
 
-        $this->config = [];
         $keys = array_keys(self::DEFAULTS);
-        foreach ($keys as $key) {
-            $this->config[$key] = array_key_exists($key, $extra) && is_array($extra[$key])
-                ? array_merge(self::DEFAULTS[$key], $extra[$key])
-                : self::DEFAULTS[$key];
-        }
+        $this->config = array_reduce(
+            $keys,
+            static function (array $config, string $key) use ($extra): array {
+                if (!array_key_exists($key, $extra)) {
+                    $config[$key] = self::DEFAULTS[$key];
+                    return $config;
+                }
+
+                if (is_array($extra[$key]) && !array_is_list($extra[$key])) {
+                    $config[$key] = array_merge(self::DEFAULTS[$key], $extra[$key]);
+                    return $config;
+                }
+
+                $config[$key] = $extra[$key];
+                return $config;
+            },
+            []
+        );
```

Comment:
Why this?

With the existing code, if some env-config configuration is there, you could end-up having:

```php
$config[ENV_CONFIGS_KEY] = array_merge(
   [
      'all',
      'development',
      'staging',
      'production',
   ],
   [
      'all',
      'development',
      'staging',
      'production',
      'something-else',
   ]
);
```

Then, below, the `envConfigs()`, method could do an `array_unique()` on the values.

This means the default envs will be always included in the list, and you can actually only _add_ more environment to the default. But in the end, the files that will be copied are the files that exists. If you don't want to have a configuration files for default environment you could just not create them.

Reply:
Idea with `array_reduce` is also that it could be possible to have values like string for possible new settings in future, because currently in case when value is not an array then default value always will be used. However, it makes sense to add `array_unique()` below the `envConfigs()` anyways.

What do you think?

Reply:
That make sense. The thing I'm not really convinced of is `array_is_list()` check. The idea that list are always not merged but always _replace_ defaults, looks pretty arbitrary to me.

So mayb we leave things are they are and _if_ and _when_ we have some string value we think about it again?

Reply:
Yes, makes sense.